### DISCUSSION TITLE
途中のアルゴリズムの変更

### DISCUSSION THREAD

**Discussion by nim-hrkn** [Tomoki-YAMASHITA/CrySPY#37] [General]
最適化構造：完全に最適化された構造と、少しだけ最適化された構造とがある。

CrySPYに採用されているアルゴリズム：
RS:　初期構造を完全に最適化する。
EA:　初期構造を完全に最適化する。親構造を作り完全に最適化する。繰り返し
BO:　初期構造から幾つか選択して完全に最適化する。回帰をして次に最適化する構造を選ぶ。繰り返し。（多数の初期構造から最適化する構造を逐次選択する方法）
LAQA:　初期構造を作る。全初期構造を少しだけ構造最適化しての全エネルギー、力を評価して次に最適化する構造を選ぶ。繰り返し。（多数の初期構造から最適化する構造を逐次選択する方法）

つくりは、

全（構造、エネルギー）、もしくは空

1a. アルゴリズム選択と
1b.初期構造生成
1c. optional.構造最適化
1d. エネルギー観測

→全（構造、エネルギー）に加える

2a. アルゴリズム選択と
2b.初期構造生成
2c. optional.構造最適化
2d. エネルギー観測

→全（構造、エネルギー）に加える

繰り返し

今は途中のアルゴリズムの変更が難しく作ってありますが、途中でアルゴリズムを変更できるようにして、
A. 断熱近似のMDとカーパリMDとの違いのように、LAQAなど完全に最適化しないで、途中の全（構造、エネルギー）を作って、全（構造、エネルギー）を作成していって、哀愁的に少しづつ最適化していく。
B. 途中で構造を加える。
C. 各ステップでEAの親構造やBOの観測済み構造とそのエネルギーがあれば、
前のアルゴリズムと異なるアルゴリズムEA,BOなどで次構造を生成する。
などが、簡単にできると良いと思います。

C. 最適化構造とrsltから次の構造を作るので変更自体は容易と思います。
例えば、EAではrsltで前世代のみを選択してから次の世代を作る関数になっているので、前世代のみを選択という部分を外に出して対応可能でしょう。

**Comment by Tomoki-YAMASHITA** [comment]
確かにRS --> BO みたいなのは特に問題なくできそうですね．出力ファイル等のスタイルを揃える必要はありそうですが．

現状 オプションの　`append_struc_ea = True`にすることで
- RS --> （EAで構造追加） --> RS 継続
- BO --> （EAで構造追加） --> BO 継続
は可能です．

LAQAの場合途中でEAによる構造追加は対応していないので
ここは少し考えないといけません．
最適化途中のエネルギーでランキングつけてEAを行うのか，フォースを含めたスコアでランキングをつけるのか．

**Reply by nim-hrkn** [reply]
機械学習古典ポテンシャルのように構造安定化途中の構造，（力），全エネルギーもガウス過程回帰のデータとして使えます．BOのデータは数を増やせます．

**Reply by Tomoki-YAMASHITA** [reply]
確かにそうですね．これはずっとやろうとは思ってるのですが，他のこともあって全然実装できていません．

**Reply by nim-hrkn** [reply]
別の説明変数の実装を含めて構造安定化途中のデータを含めた実装できればと思います．

書き忘れていましたが，aiida-espressoやaiida-vaspだとaiidaに計算サーバーの実行ディレクトリを管理させて，計算サーバーに電荷やポテンシャルなどのサイズが大きいDFTの再計算ファイルをおいたままにして，再計算時に計算サーバーにおいたファイルを利用した再計算が可能に実装されています．AiiDA-coreにそういう機能があり，私が作ったaiida-akaikkrもそうなっています．
手動によるローカルホスト↔計算サーバーのファイル移動が不要というだけですが，LAQAや少しづつ計算させる場合はローカルホストディレクトリと計算サーバーディレクトリ管理の手間が減ります．

**Reply by Tomoki-YAMASHITA** [reply]
LAQAなんかと相性が良いわけですね．

**Comment by nim-hrkn** [comment]
途中で全てのアルゴリズムを変えるのは原理的には外部構造を読み込むinitializeをやり直すとできるのですね。
Changing the algorithms in the middle of the process can in principle be done by restarting with structures already generated, isn't it?  

現在は構造集合に対してIDを使って制御しています。しかし、その方法はせっかくノードをグラフで可視化できる仕組みを使っているのにうまく使っていません。構造を一つづつ繋げて、実績グラフ上でEAの親が明確に分かる形にしたいのですが、それはプログラム構造をかなり変えるので時間がかかります。
Currently, ID is used to control for the structure set. However, that method is not a good use of the mechanism that allows nodes to be visualized in a graph. I would like to connect the structures to one another, so that the parents in the EA are elucidated on the provenance graph, but that takes much time because it changes the program structure quite a bit.

**Reply by yw-fang** [reply]
```
途中で全てのアルゴリズムを変えるのは原理的には外部構造を読み込むinitializeをやり直すとできるのですね。
Changing the algorithms in the middle of the process can in principle be done by restarting with structures already generated, isn't it?
```
In principle, we indeed can change the algorithms during the crystal structure prediction However, one generally do not really do like this because it means the strategy to locate the potential energy surface is also changed. Such a mixing is good or not is still unknown to us. 
Aas said by @Tomoki-YAMASHITA, CrySPY allows us to load already optimized structures as **initial** guess. To my experience (of CrySPY and CALYPSO), I have ever used PyXtal to generate many random structures and load their optimized ones to the former two code as the first generation.

**Reply by yw-fang** [reply]
```
現在は構造集合に対してIDを使って制御しています。しかし、その方法はせっかくノードをグラフで可視化できる仕組みを使っているのにうまく使っていません。構造を一つづつ繋げて、実績グラフ上でEAの親が明確に分かる形にしたいのですが、それはプログラム構造をかなり変えるので時間がかかります。
Currently, ID is used to control for the structure set. However, that method is not a good use of the mechanism that allows nodes to be visualized in a graph. I would like to connect the structures to one another, so that the parents in the EA are elucidated on the provenance graph, but that takes much time because it changes the program structure quite a bit.
```
When using CrySPY as a standalone code, using "ID" to label the structures is natural, which is same as USPEX and CALYPSO.
I understand your point that aiida-core wants to use a node to point to a structure, in which CrySPY-ID means nothing to the workflow.
When developing aiida-cryspy, probably you can stick a "node" to a structure using the logic of aiida-core, but in the meantime, we can still keep the pkl data with preserving the original CrySPY-ID for each structure.

**Reply by nim-hrkn** [reply]
現在、aiida_cryspyはRS, EA, LAQA, BOとそれらの繰り返しが使えるようになっています。途中でアルゴリズムを変えたかったのですが、id_dataや(rs,ea,laqa,bo)_dataがアルゴリズムにより異なることが障害になっていました。この変更は面倒です。しかし、すでに計算した構造ファイルを読み込むことでaiida_cryspyでもコードを書き換えることなく途中での可能になります。
Currently, RS, EA, LAQA, BO and their repetitions are available in aida_cryspy. Furthermore, I wanted to change the algorithms in the process, but the obstacle was that the id_data and (rs,ea,laqa,bo)_data were different for different algorithms. This change is tedious. However, by loading structures that has already been calculated, aiida_cryspy can also change the algorithms in the middle of the process without rewriting the code.

**Reply by Tomoki-YAMASHITA** [reply]
なるほど．

**Reply by nim-hrkn** [reply]
Note that aiida_cryspy writes the pkl data, but doesn't read them. The data in the pkl data can be obtained from nodes of workflows in AiiDA.

**Comment by Tomoki-YAMASHITA** [comment]
> Changing the algorithms in the middle of the process can in principle be done by restarting with structures already generated, isn't it?

できますが，外部構造は初期構造として読み込まれます．最適化済みの構造も初期構造として読み込むことができますが，その場合はそれらを初期構造とした構造最適化計算がもう一度行われてしまいます. 最適化済み構造とそのエネルギーのセットを読み込むようなオプションはまだ作っていません
Yes, but CrySPY can load already generated structures as initial structures.  
You can also load already optimized structures as initial structures, in which case structure optimization calculation (useless) will be performed again.
I have not yet made an option to load a set of optimized structures and their energies.

**Comment by Tomoki-YAMASHITA** [comment]
> Currently, ID is used to control for the structure set. However, that method is not a good use of the mechanism that allows nodes to be visualized in a graph. I would like to connect the structures to one another, so that the parents in the EA are elucidated on the provenance graph, but that takes much time because it changes the program structure quite a bit.

一応EA_originというファイルは出力させていますが，これが利用できるかどうかはわかりません．
出力元のデータはEA_data.pklに入っています．

The file EA_origin is output, but I am not sure if it helps you or not.
The output data is contained in EA_data.pkl.


example of EA_origin
```
Gen  Struc_ID    Operation    Parent
   1         0       random      None
   1         1       random      None
   1         2       random      None
   1         3       random      None
   1         4       random      None
   1         5       random      None
   1         6       random      None
   1         7       random      None
   1         8       random      None
   1         9       random      None
   1        10       random      None
   1        11       random      None
   1        12       random      None
   1        13       random      None
   1        14       random      None
   1        15       random      None
   1        16       random      None
   1        17       random      None
   1        18       random      None
   1        19       random      None
   2        20    crossover   (15, 9)
   2        21    crossover    (9, 1)
   2        22    crossover   (9, 10)
   2        23    crossover   (9, 19)
   2        24    crossover   (9, 10)
   2        25    crossover   (17, 6)
   2        26    crossover    (1, 9)
   2        27    crossover  (19, 15)
   2        28    crossover  (19, 15)
   2        29    crossover    (9, 1)
   2        30  permutation     (13,)
   2        31  permutation      (6,)
   2        32  permutation      (2,)
   2        33       strain      (1,)
   2        34       strain     (13,)
   2        35       strain      (6,)
   2        36       strain      (1,)
   2        37       strain      (1,)
   2        38       random      None
   2        39       random      None
   2         9        elite     elite
   2         1        elite     elite
```

**Reply by nim-hrkn** [reply]
まず、現在のCrySPYではなく、私はAiiDA実装について書いていました。更に、幾つかのことを混ぜて話していて理解しにくくて申し訳ありません。
この返信ではEA時の関係について説明します。CrysPyだと構造間の関係はEA_originで別に記述し、IDを用いて両者を紐づけします。yw-fangさんが書かれた通り、IDを用いて紐づけすることが一般的なやり方と思います。一方、再びyw-fangさんが書かれた通り、AiiDAでは、構造をノードとしてimplementできると入力ノードー変換装置ー出力ノードとの関係を直接示すことができます。添付するような図はCrysPyでは EA_originを可視化しただけですが、AiiDAで構造をノードとして書くことができると、実績グラフとして同じ図を示すことができます。また、AiiDAではこの図は計算結果を示す図ではなく、計算過程を示す図となります。
First, I was writing about the AiiDA implementation, not the current CrySPY. Furthermore, I apologize for mixing several things up and making it difficult to understand.
This reply describes the relationship during EA. In CrySPY, the relationship between structures is described separately in EA_origin and IDs which define the structures, and the IDs are used to link the two. As Dr. yw-fang wrote, IDs are a common way to identify the structures. On the other hand, as Dr. yw-fang wrote again, AiiDA can directly show the relationship among input nodes, conversion devices, and output nodes when the structures are implemented as nodes. The attached figure is just a visualization of EA_origin in CrySPY, but if the structures can be written as nodes in AiiDA, the same figure can be shown as a provenance graph. Furthermore, in AiiDA, it is not a diagram showing the result of the calculation, but a diagram showing the process of the calculation.

![進化論アルゴリズム](https://user-images.githubusercontent.com/70899/194446461-07f2e69e-02e2-46f0-aee3-efdee6369f6c.png)

**Reply by nim-hrkn** [reply]
このノード（構造）間の関係を直接理解できる実装をするにはCrySPYはとても良い例です。しかし、これはCrySPYを完全に書き直すことになり時間がかかるので後回しにします。
CrySPY is a very good example of how to implement a direct understanding of the relationships between these nodes (structures). However, this would require a complete rewrite of CrySPY, which would be time-consuming, so it will be done later. (It may be done.)

**Comment by nim-hrkn** [comment]
This is a summary of discussions
Structure optimizations are carried out in structure exploration programs.
![initial_final_structures](https://user-images.githubusercontent.com/70899/194470961-8057fa9b-ffbf-4e2e-8a14-72b86916a34e.png)
BO currently uses final structures ($R_{ID}^{final}$) and enthalpies (E_{ID}^{final}). LAQA is another method of structure exploration which uses $R_{ID}^{i}$ and $acq_{ID}^{i}$ during optimizaton process using  enthalpies, forces and stresses.
![laqa](https://user-images.githubusercontent.com/70899/194471155-5f1da62c-ef51-4b88-bbad-aa33a94a1e6d.png)
 $R_{ID}^{i}$ and $acq_{ID}^{i}$ can be used also in BO to increase data. 
The typical $E_{ID}^{final}$ curve is as follows.
![Evsiteration](https://user-images.githubusercontent.com/70899/194471565-ed0ad065-aefe-4af2-a750-9ada57d441d3.png)
(a) and (b) are almost the same structures, but it takes times to be converged.  (This is the same as when the optimization threshold is loosened.)
Apparently, it is unnecessary to restrict data only within the final structures. The data during optimization can be used for BO as well as EA, though one must take into account whether the structures (circles in the figure) are the same or not. 
The key is to increase diversity. As a program that uses BO and LAQA, it is a natural extension to use data in the process of optimization. I would like the aiida_cryspy package to be flexible enough to easily apply these algorithims.

**Reply by Tomoki-YAMASHITA** [reply]
Thanks for the good summary!
